### PR TITLE
fix(train): honor chat template overrides across trainers

### LINKED ISSUE #804: data.chat_template is applied by SFT and distill only: DPO renders conversational rows with the model's shipped template instead
## The defect

`data.chat_template` ("Override the tokenizer chat template", `config/schema.py:398`) is installed on the tokenizer by `apply_chat_template_override`, which has one caller: `build_format_row` (`src/soup_cli/data/sft_format.py:52`). `build_format_row` has two callers, `trainer/sft.py:729` and `trainer/distill.py:469`. No other trainer applies the override, and no validator restricts the field by task.

The preference trainers hand message-list rows to TRL, which renders them with `tokenizer.chat_template`, the shipped one. GRPO and Online DPO go further: when the tokenizer has no template they install a hardcoded fallback (`trainer/grpo.py:360-363`, `trainer/online_dpo.py:353-354`), never the user's.

## Reproduction

Tiny random-init Llama on CPU carrying SmolLM2's tokenizer (which ships a ChatML template with a default system prompt). `data.chat_template` set to a custom Jinja string that emits a `[[SOUP-OVERRIDE <role>]]` marker. Same override, same prompt, two tasks:

```

config accepted: task=dpo data.chat_template set = True

columns: ['prompt', 'chosen', 'rejected', 'prompt_ids', 'chosen_ids', 'rejected_ids']

DPO rendered prompt: '<|im_start|>system\nYou are a helpful AI assistant named SmolLM, trained by Hugging Face<|im_end|>\n<|im_start|>user\nWhat is 2+2?<|im_end|>\n<|im_start|>assistant\n4<|im_end|>\n'

tokenizer template contains override marker: False

SFT rendered row: '[[SOUP-OVERRIDE user]] What is 2+2? [[SOUP-OVERRIDE assistant]] 4 '

```

(DPO rows: `{"prompt": [{"role": "user", ...}], "chosen": [{"role": "assistant", ...}], "rejected": [...]}`, the conversational shape `_convert_dpo` passes through, `data/formats.py:245-259`.) The DPO run trains on the shipped template, including a system prompt the user never wrote.

Acceptance by task, `load_config_from_string` with `data.chat_template: chatml`:

```

ACCEPT dpo  ACCEPT kto  ACCEPT orpo  ACCEPT simpo  ACCEPT ipo  ACCEPT bco

ACCEPT reward_model  ACCEPT grpo  ACCEPT online_dpo  ACCEPT pretrain

ACCEPT embedding  ACCEPT classifier  ACCEPT ppo

```

I traced but did not execute KTO / ORPO / SimPO / IPO / BCO / reward_model: none of them references `chat_template`, `build_format_row` or `apply_chat_template_override`, so on those the field is accepted and never read, and any conversational rendering TRL does uses the shipped template, as measured for DPO.

## Why it matters

The common reason to set the field is SFT-then-DPO on one base: the SFT stage is trained on the override, and the DPO stage on the same config silently switches to another format. The model is then preference-tuned on a prompt format it was never instruction-tuned on, and inference tools that read the saved tokenizer see whichever template that trainer left behind. For a base model with no template, SFT honours `chat_template: chatml` while GRPO renders prompts as bare `content\n` lines.

This is the same failure #332 fixed for `soup serve --backend vllm` (a hand-rolled prompt format a model was never trained on); here it happens between two training stages.

## Proposed fix

Apply the override once, where the tokenizer is loaded, in every trainer: call `apply_chat_template_override(self.tokenizer, cfg.data.chat_template, console=console)` right after `AutoTokenizer.from_pretrained` (a shared tokenizer-loading helper would stop this recurring). GRPO / Online DPO then install their fallback only if the tokenizer *still* has no template. Tasks that never render chat (pretrain, embedding, classifier) should reject the field at load with the task named rather than accept it.

Also check that the saved tokenizer in the output dir carries the override, so `soup chat` / `soup serve` render what training used.

## Acceptance criteria

- [ ] A test that builds the DPO trainer on a tiny CPU model with a marker template and asserts the marker appears in the decoded `prompt_ids`. It fails on current code.

- [ ] The same assertion for KTO, ORPO, SimPO, IPO, BCO, reward_model, GRPO and Online DPO (parametrized).

- [ ] With `chat_template` set on a template-less tokenizer, GRPO renders with it, not with the `{{ msg['content'] }}` fallback.

- [ ] `data.chat_template` on pretrain / embedding / classifier either does something or is refused at load.

---

**Related, filed the same day:**

- #806 — use_flash_attn and use_liger are read only by SFT's text path, yet validated on every task and switched on ...

### PR DESCRIPTION
**Pull request by Amix29**
Closes #804.

## What does this PR do?

`data.chat_template` is now installed on the tokenizer before TRL renders conversational data in DPO, KTO, ORPO, SimPO, IPO, BCO, Reward Model, GRPO, and Online DPO.

- The override runs after whichever tokenizer load path was selected, so ordinary Transformers, Unsloth, and layer-streaming preference paths share the same behavior.
- GRPO and Online DPO only install their built-in fallback if the configured override still leaves the tokenizer without a template.
- The tokenizer object later saved by each trainer carries the same template used for training.
- `pretrain`, `embedding`, `classifier`, `reranker`, and `cross_encoder` now reject `data.chat_template` with the task named instead of accepting a field they never use.
- The chat-template documentation states the supported scope.

## Regression coverage

`tests/test_issue804_chat_template_trainers.py` includes:

- a real tiny CPU DPO setup whose prepared `prompt_ids` decode with an `OVERRIDE` marker and not the model's shipped `SHIPPED` marker;
- save/reload proof that the tokenizer persists the override;
- runtime wiring checks for the eight offline conversational wrappers;
- explicit ordering checks for the GRPO and Online DPO fallbacks;
- a real local Online DPO tokenizer/model load showing the override beats its fallback;
- task-named refusal tests for all five non-chat tasks.

The existing multi-task fixture was adjusted so it no longer supplies a now-invalid chat template to non-chat tasks.

## Verification

- `ruff check src/soup_cli/ tests/` — clean
- issue-specific + existing chat-template tests — 46 passed
- live TRL preference contracts + multi-task setup matrix — 131 passed
- config + MLX tests — 31 passed
- broad suite before stopping after the first unrelated failure — 2,920 passed, 12 skipped, 5 deselected
- the unrelated failure reproduces alone on `origin/main`-independent code: `tests/test_issue204_langfuse_pull.py::test_control_the_leak_scan_sees_a_token_rich_highlighted_and_folded` assumes Rich will wrap one planted string, but the installed Rich renders both unbroken

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Refactor / cleanup
- [x] Documentation
- [x] Tests

## Checklist

- [x] `ruff check src/soup_cli/ scripts/ tests/` passes
- [ ] `pytest tests/ -v` passes — one unrelated, isolated Rich wrapping assertion fails as recorded above
- [x] Updated the relevant page under `docs/` (README does not mention this field)
- [x] Added a changelog fragment, or this change has no user-visible impact

### CODE REVIEW

**Comment by MakazhanAlpamys** [comment]
Reviewed at `21ca2c86`. I checked the thing this repo keeps getting wrong: whether the field is actually **read**, or only plumbed somewhere plausible-looking.

**It is read.** `apply_chat_template_override` mutates the tokenizer in place, the call sits *after* the three-way dispatch (streaming / unsloth / transformers) so all three load paths are covered by construction, and all nine trainers hand **that same object** to TRL — `processing_class=self.tokenizer` at `dpo:222 kto:211 orpo:222 simpo:222 ipo:180 bco:234 reward_model:175 grpo:526 online_dpo:308`, plus `grpo:416/559` — each saving it exactly once. Your DPO test closes the loop end-to-end on decoded `prompt_ids`, and the `"SHIPPED" not in rendered` assertion is what makes it a real test rather than a presence check.

**Per-trainer coverage confirmed by mutation, which was my main worry.** I removed the override from **`simpo` alone** — the pre-PR state for one trainer — and got exactly **2 failures, both parametrized on `simpo`**, with the other eight green. So a single-trainer regression is visible rather than hidden behind one shared assertion. I also kept the call and its spelling but passed `None`, which is the "validated, documented, read by nothing" shape this project is named for, and **9 tests fail** including the DPO end-to-end one — while the `inspect.getsource` check happily passes. The behavioural half is doing the work. Making the GRPO/Online-DPO fallbacks unconditional fails 4 (2 behaviourally); disabling the non-chat rejection fails 5, one per task. No survivors.

**Runs:** your file 28 passed; +chat-template/config/contract files 451 passed; +21 files importing the nine changed trainers 580 passed, 0 failed; the #748 / #755 / #487 / #775 repo guards 106 passed; ruff clean. No shipped config is affected — `chat_template` appears in zero recipes, templates or examples, so none of the 167 recipes can trip the new validator.

I was suspicious of the `and not self.training.stream_layers` clause at `schema.py:4293` — it looks like a guard written to satisfy its own test — so I executed it. All five unsupported tasks are **still refused** with `stream_layers: true`, just by the streaming validator's message instead. `unsupported ∩ stream-capable = ∅`, so it only decides *which* of two errors the user sees and no config leaks through. **Please add a one-line comment saying that is why it is there**, or someone will delete it, or worse, copy the pattern.

**One test-file adaptation I checked line by line:** `tests/test_issue353_seed_every_task.py` stopped feeding `chat_template: chatml` to `pretrain`/`embedding`/`classifier`, because those configs are now invalid **by this PR's own validator**. No assertion weakened; the fixture simply had to stop constructing configs that no longer exist.

**Follow-up, please file it.** `ppo`, `prm`, `asr` (and `moe_lora_routing`, `unlearn`) still **accept `data.chat_template` and read nothing** — I enumerated all 23 tasks against the real `SoupConfig`, and `grep` finds no hits in any of the five. `ppo` is the notable one: `ppo.py:914-920` hand-joins `msg["content"]` with spaces, the same class of bug #332 fixed for vLLM. Outside #804's acceptance criteria so not this PR's job, but the class is not closed.

**Two smaller things:**

- `test_grpo_applies_override_before_fallback` and its online-DPO sibling assert source **order**, and I measured that both orders are behaviourally identical — the override wins either way, because it overwrites. They also match a literal adjacent line, so a harmless refactor breaks them, and your behavioural tests already cover the real hazard. `CLAUDE.md`'s own warning about spelling-versus-behaviour tests applies; consider dropping them.
- The fragment needs `by @Amix29`. The checklist also ticks "Updated README.md" while README is unchanged — nothing was owed there, it never mentioned the field, so that is just an inaccurate box.

**Merge-order heads-up:** you and #908 both add exactly one line at the *same* import anchor in all nine of these files, so that is a guaranteed add/add conflict ×9 for whoever lands second — trivial to resolve, but re-run `ruff check` afterwards since import order is enforced. Your own #923 does **not** conflict with this one (different functions; its only import change there is function-local).

Approving once the remaining CI cells report.

**Comment by Amix29** [comment]
Addressed the review at `c55764c`:

- documented why the chat-template validator defers to the streaming task validator
- removed the two fragile source-order assertions; the behavioral GRPO/Online-DPO tests remain
- added the required changelog attribution
- corrected the PR checklist wording
- filed #936 for the remaining PPO/PRM/ASR/MoE/unlearn accepted-but-unused paths

Validation: 43 targeted tests passed and the full Ruff scope is clean. Rebased onto current `origin/main`.

**Comment by MakazhanAlpamys** [comment]
Verified at `c55764c`. All five items done, and I checked each rather than taking the summary.

- The `stream_layers` clause now carries its reason inline (`schema.py:4294-4295`), which is what stops the next reader deleting it or copying the pattern.
- Both source-order assertions are gone — `grep -c "applies_override_before_fallback"` returns **0** — and the behavioural GRPO and Online-DPO tests remain, which are the ones that actually catch a fallback overwriting the user's override.
- The fragment reads `(#804 by @Amix29 in #919)`.
- #936 is filed for the `ppo` / `prm` / `asr` / `moe_lora_routing` / `unlearn` accepted-but-unread paths, which keeps that class open and named rather than buried in a review thread.

Removing tests you had written, because they assert a spelling rather than a behaviour, is a harder thing to do in review than adding more — and it is the right call here. `CLAUDE.md` has the same lesson written down from #357/#396, and it keeps costing this project releases.

Merging as soon as the checks report — they are still queued on this head.

**Ordering, mine to manage:** you and #908 both add exactly one line at the *same* import anchor in all nine of these trainer files, so whoever lands second gets nine add/add conflicts. They are trivial — keep both lines, alphabetise — but re-run `ruff check` afterwards, since import order is enforced. Your own #923 does not conflict with this one.

**Comment by MakazhanAlpamys** [comment]
Merged as `c6b61657`. Thank you — and the review round on this one is the part I want on the record.

You had two assertions on **source order** (`applies_override_before_fallback`), and you deleted them when I asked rather than defending them. That matters because an assertion that one statement precedes another catches the refactor that moves it, not the defect it is aimed at — a fallback overwriting the user's override. The behavioural GRPO and Online-DPO tests that replaced them do catch it. `grep -c "applies_override_before_fallback"` is now 0.

Putting the reason for the `stream_layers` clause inline (`schema.py:4294-4295`) is the other thing that will still be paying off in a year: it is what stops the next reader deleting it or copying the pattern without knowing why it exists.

And you filed #936 for the four trainers still not reading the field rather than widening this PR to reach them. Right call.

Your CONTRIBUTORS.md entry is in and #804 is closed.
